src/interface_types/resource_handles.rs

```diff
@@ -0,0 +1,371 @@
+//////////////////////////////////////////////////////////////////////////////////
+/// Clear
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Clear {
+    Owner,
+    Platform,
+}
+
+impl From<Clear> for AuthHandle {
+    fn from(clear: Clear) -> AuthHandle {
+        match clear {
+            Clear::Owner => AuthHandle::OwnerHandle,
+            Clear::Platform => AuthHandle::PlatformHandle,
+        }
+    }
+}
+
+impl TryFrom<AuthHandle> for Clear {
+    type Error = Error;
+
+    fn try_from(auth_handle: AuthHandle) -> Result<Self> {
+        match auth_handle {
+            AuthHandle::OwnerHandle => Ok(Clear::Owner),
+            AuthHandle::PlatformHandle => Ok(Clear::Platform),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
+        }
+    }
+}
+//////////////////////////////////////////////////////////////////////////////////
+/// NvAuth
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum NvAuth {
+    Platform,
+    Owner,
+    NvIndex(NvIndexHandle),
+}
+
+impl From<NvAuth> for AuthHandle {
+    fn from(nv_auth: NvAuth) -> AuthHandle {
+        match nv_auth {
+            NvAuth::Platform => AuthHandle::PlatformHandle,
+            NvAuth::Owner => AuthHandle::OwnerHandle,
+            NvAuth::NvIndex(nv_index_handle) => nv_index_handle.into(),
+        }
+    }
+}
+
+impl TryFrom<AuthHandle> for NvAuth {
+    type Error = Error;
+
+    fn try_from(auth_handle: AuthHandle) -> Result<NvAuth> {
+        match auth_handle {
+            AuthHandle::PlatformHandle => Ok(NvAuth::Platform),
+            AuthHandle::OwnerHandle => Ok(NvAuth::Owner),
+            _ => Ok(NvAuth::NvIndex(NvIndexHandle::from(auth_handle))),
+        }
+    }
+}
+//////////////////////////////////////////////////////////////////////////////////
+/// Lockout
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Lockout {
+    Lockout,
+}
+
+impl From<Lockout> for TpmConstantsHandle {
+    fn from(_: Lockout) -> TpmConstantsHandle {
+        TpmConstantsHandle::Lockout
+    }
+}
+
+impl TryFrom<TpmConstantsHandle> for Lockout {
+    type Error = Error;
+
+    fn try_from(tpm_constants_handle: TpmConstantsHandle) -> Result<Lockout> {
+        match tpm_constants_handle {
+            TpmConstantsHandle::Lockout => Ok(Lockout::Lockout),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
+        }
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////////
+// AttachedComponent
+//
+// TODO: Figure out how to implement this and how this should relate
+//       AttachedComponent enum value in crate::handles::tpm::TpmHandle.
+//////////////////////////////////////////////////////////////////////////////////
```

Comment:
The most sensible solution seems to be implementing `AC_FIRST:AC_LAST` as wrappers around a numeric value and checking that the value is in the correct range when constructing. It seems the type for `NV_INDEX_FIRST:NV_INDEX_LAST` (which I assume is `NvIndexHandle`) doesn't have checks for what values are valid or not.

Reply:
No the reason for this is because those values have been commented out in the constants file:

```src/constants/tss.rs```
```rust
// the following constants cannot be instantiated since the TPM2_HT constants are u8
//pub const TPM2_HR_PCR               : TPM2_HC = TPM2_HT_PCR << TPM2_HR_SHIFT;
//pub const TPM2_HR_HMAC_SESSION      : TPM2_HC = TPM2_HT_HMAC_SESSION << TPM2_HR_SHIFT;
//pub const TPM2_HR_POLICY_SESSION    : TPM2_HC = TPM2_HT_POLICY_SESSION << TPM2_HR_SHIFT;
//pub const TPM2_HR_TRANSIENT         : TPM2_HC = TPM2_HT_TRANSIENT << TPM2_HR_SHIFT;
//pub const TPM2_HR_PERSISTENT        : TPM2_HC = TPM2_HT_PERSISTENT << TPM2_HR_SHIFT;
//pub const TPM2_HR_NV_INDEX          : TPM2_HC = TPM2_HT_NV_INDEX << TPM2_HR_SHIFT;
//pub const TPM2_HR_PERMANENT         : TPM2_HC = TPM2_HT_PERMANENT << TPM2_HR_SHIFT;
//pub const TPM2_PCR_FIRST            : TPM2_HC = TPM2_HR_PCR + 0; /* first PCR */
//pub const TPM2_PCR_LAST             : TPM2_HC = TPM2_PCR_FIRST + TPM2_MAX_PCRS - 1; /* last PCR */
//pub const TPM2_HMAC_SESSION_FIRST   : TPM2_HC = TPM2_HR_HMAC_SESSION + 0; /* first HMAC session */
//pub const TPM2_HMAC_SESSION_LAST    : TPM2_HC = TPM2_HMAC_SESSION_FIRST + 0x00fffffe; /* last HMAC session */
//pub const TPM2_LOADED_SESSION_FIRST : TPM2_HC = TPM2_HMAC_SESSION_FIRST; /* used in GetCapability */
//pub const TPM2_LOADED_SESSION_LAST  : TPM2_HC = TPM2_HMAC_SESSION_LAST; /* used in GetCapability */
//pub const TPM2_POLICY_SESSION_FIRST : TPM2_HC = TPM2_HR_POLICY_SESSION + 0; /* first policy session */
//pub const TPM2_POLICY_SESSION_LAST  : TPM2_HC = TPM2_POLICY_SESSION_FIRST + 0x00fffffe; /* last policy session */
//pub const TPM2_TRANSIENT_FIRST      : TPM2_HC = TPM2_HR_TRANSIENT + 0; /* first transient object */
//pub const TPM2_ACTIVE_SESSION_FIRST : TPM2_HC = TPM2_POLICY_SESSION_FIRST; /* used in GetCapability */
//pub const TPM2_ACTIVE_SESSION_LAST  : TPM2_HC = TPM2_POLICY_SESSION_LAST; /* used in GetCapability */
//pub const TPM2_TRANSIENT_LAST       : TPM2_HC = TPM2_TRANSIENT_FIRST + 0x00fffffe; /* last transient object */
//pub const TPM2_PERSISTENT_FIRST     : TPM2_HC = TPM2_HR_PERSISTENT + 0; /* first persistent object */
//pub const TPM2_PERSISTENT_LAST      : TPM2_HC = TPM2_PERSISTENT_FIRST + 0x00FFFFFF; /* last persistent object */
//pub const TPM2_PLATFORM_PERSISTENT  : TPM2_HC = TPM2_PERSISTENT_FIRST + 0x00800000; /* first platform persistent object */
//pub const TPM2_NV_INDEX_FIRST       : TPM2_HC = TPM2_HR_NV_INDEX + 0; /* first allowed NV Index */
//pub const TPM2_NV_INDEX_LAST        : TPM2_HC = TPM2_NV_INDEX_FIRST + 0x00FFFFFF; /* last allowed NV Index */
```